### PR TITLE
Avoid `stopifnot()` in `new_list_of()`

### PR DESCRIPTION
**Pull request by DavisVaughan**
Part of the slowness of `dplyr:::new_rowwise_df()`

### CODE REVIEW

File: R/type-list-of.R [hunk 1]
```diff
@@ -59,19 +59,28 @@ as_list_of.list <- function(x, ..., .ptype = NULL) {
 #' @keywords internal
 #' @export
 new_list_of <- function(x = list(), ptype = logical(), ..., class = character()) {
-  stopifnot(is.list(x))
-  stopifnot(vec_size(ptype) == 0)
+  if (!vec_is_list(x)) {
+    abort("`x` must be a list")
```
Comment (lionel-) [inline R/type-list-of.R]:
```suggestion
    abort("`x` must be a list.")
```

File: R/type-list-of.R [hunk 2]
```diff
@@ -59,19 +59,28 @@ as_list_of.list <- function(x, ..., .ptype = NULL) {
...
+  }
+
+  if (vec_size(ptype) != 0L) {
+    abort("`ptype` must have size 0")
```
Comment (lionel-) [inline R/type-list-of.R]:
```suggestion
    abort("`ptype` must have size 0.")
```

File: R/type-list-of.R [hunk 3]
```diff
@@ -59,19 +59,28 @@ as_list_of.list <- function(x, ..., .ptype = NULL) {
...
 
   new_vctr(x, ..., ptype = ptype, class = c(class, "vctrs_list_of"))
 }
 
 #' @export
 #' @rdname list_of
 validate_list_of <- function(x) {
-  stopifnot(is.list(x))
+  if (!vec_is_list(x)) {
+    abort("`x` must be a list")
```
Comment (lionel-) [inline R/type-list-of.R]:
```suggestion
    abort("`x` must be a list.")
```

File: R/type-list-of.R [hunk 4]
```diff
@@ -59,19 +59,28 @@ as_list_of.list <- function(x, ..., .ptype = NULL) {
...
+  }
 
   ptype <- attr(x, "ptype")
-  stopifnot(vec_size(ptype) == 0)
+  if (vec_size(ptype) != 0L) {
+    abort("`ptype` must have size 0")
```
Comment (lionel-) [inline R/type-list-of.R]:
```suggestion
    abort("`ptype` must have size 0.")
```

File: src/type-list-of.c [hunk 5]
```diff
@@ -3,7 +3,7 @@
 
 // [[ include("utils.h") ]]
 SEXP new_list_of(SEXP x, SEXP ptype) {
-  if (TYPEOF(x) != VECSXP) {
+  if (!vec_is_list(x)) {
```
Comment (lionel-) [inline src/type-list-of.c]: This explicit support for S3 lists makes me wonder if we should zap the attributes before initialising the list-of.

Comment (lionel-) [reply src/type-list-of.c]: We probably should.

Comment (DavisVaughan) [reply src/type-list-of.c]: The R level one does, so yea probably. I'll update. 

```
attributes(vctrs::new_list_of(structure(list(), x = 1)))
#> $ptype
#> logical(0)
#> 
#> $class
#> [1] "vctrs_list_of" "vctrs_vctr"
```

**Pull request merged**